ksqldb-engine/src/main/java/io/confluent/ksql/engine/InsertValuesExecutor.java

```diff
@@ -266,12 +274,52 @@ private void throwIfDisabled(final KsqlConfig config) {
           FormatFactory.fromName(dataSource.getKsqlTopic().getKeyFormat().getFormat()),
           topicName,
           true,
+          "write",
           e);
       LOG.error("Could not serialize key.", e);
       throw new KsqlException("Could not serialize key: " + keyValue, e);
     }
   }
 
+  /**
+   * Ensures that the key schema that we generate will be identical
+   * to the schema that is registered in schema registry, if it exists.
+   * Otherwise, it is possible that we will publish messages with a new
+   * schemaID, meaning that logically identical keys might be routed to
+   * different partitions.
+   */
+  private void ensureKeySchemasMatch(
+      final PersistenceSchema keySchema,
+      final DataSource dataSource,
+      final ServiceContext serviceContext
+  ) {
+    final KeyFormat keyFormat = dataSource.getKsqlTopic().getKeyFormat();
+    final Format format = FormatFactory.fromName(keyFormat.getFormat());
+    if (!format.supportsFeature(SerdeFeature.SCHEMA_INFERENCE)) {
+      return;
+    }
+
+    final ParsedSchema schema = format
+        .getSchemaTranslator(keyFormat.getFormatInfo().getProperties())
+        .toParsedSchema(keySchema);
+
+    try {
+      final SchemaMetadata latest = serviceContext.getSchemaRegistryClient()
+          .getLatestSchemaMetadata(
+              KsqlConstants.getSRSubject(dataSource.getKafkaTopicName(), true));
+      if (!latest.getSchema().equals(schema.canonicalString())) {
```

Comment:
Are we handling the case where no schema is present in Schema Registry? I know this doesn't happen under normal circumstances since ksqlDB registers schemas at source creation time, but we should handle potential subsequent deletion by the user gracefully rather than throwing an NPE.

Reply:
I thought it throws an exception if the schema does not exist with a 404 but I will double check that (so it would be handled in the catch block below)

Reply:
It appears that your comment above handles this case 😂